### ISSUE TITLE
Unlock option only in page permission causing admin UI crash with internal error

### ISSUE THREAD

**Issue by weiwang-gsa** [wagtail/wagtail#12628]
### Issue Summary

Issue was discovered when trying to add new user group with review only access. When only assign “Unlock” option to selected page in Page permissions, and users belong to this new user group try to login to admin will have admin interface crash with internal errors. 

### Steps to Reproduce

1. Login to wagtail admin
2. Create a new user group, e.g. ReadOnlyGroup
3. Under other permissions, assign “Can access wagtail admin” option to it.
4. Under page permission, select a page to this user group and keep all option unselected and save it, it saved as if it is successful, but if you go back and check this user group settings, the page selection from the page permissions is gone. It is silent failed to save.
5. If  check “unlock” option—> save it—>page save successfully. ![Image](https://github.com/user-attachments/assets/6eb556ea-9371-41f6-a513-09b0893ab431)

6. Create a user (e.g. readonly_tester) and assign it to the ReadOnlyGroup just created
7. Login with this new user (readonly_tester)
8. Error - admin login page crashed with Internal error on server side,![Image](https://github.com/user-attachments/assets/a5c0d74a-f7d0-4edd-a7bd-6601ad3d59d7)
 locally it gave tracing error "Error during template rendering
In template site_summary_pages.html, error at line 5. Reverse for 'wagtailadmin_explore' with arguments '('',)' not found. 1 pattern(s) tried: ['admin/pages/(?P<parent_page_id>[0-9]+)/\\Z']" seems unlock option doesn't end up with a page primary key been associated.![Image](https://github.com/user-attachments/assets/f602a779-d84f-412b-9be5-150eda6128fe)

### Technical details

* Python version: 3.12.5
* Django version: 5.1
* Wagtail version: 6.2
* Browser version: Chrome 131

### Requested features:

* Bug fix for internal error when only select unlock options
* Review only access to allow user group users just review the page set up but has no ability to add or edit pages.

**Comment by zerolab** [comment]
Confirmed. The quick and dirty fix would be to update the template to conditionally generate the URL.
However there are other issues as you explore. The proper fix here is to consider what it means to only have the "unlock" permission as the main drivers for the permission checks are the add/edit/delete permissions.

**Comment by weiwang-gsa** [comment]
@zerolab Thank you for quick response, our original purpose is to have a **"read only access" or "comment only"user group, the users only can view and comment on the page content only, but can't add/edit/delete the page**. We tried just add the page to the page permission section without selecting any of existing options (add/edit/delete/lock/unlock) and hope that kinda give the read only ( comment only) access, but without select any of options, the page won't be saved to page permission, so the user will not able to see the page at all. so I tried to see if I only select "unlock" a page, if this will give close to ready only (comment only) access, and then found this bug.

**Comment by zerolab** [comment]
This particular requirement has come up a few times (including an active project for me). One way to allow the "comment only" group is have a workflow task limited to them. The one catch is that in order to save the comments one needs to save the draft and that allows changes at the moment.

Have you seen/tried https://github.com/wagtail-nest/wagtail-review? It could be enough for your use case

It is worth mentioning that the [roadmap](https://wagtail.org/roadmap/) has workflow enhancements as a future item, so use cases and issues like this are welcome (not to mention the dev power, collaboration or sponsorship)

**Comment by weiwang-gsa** [comment]
@zerolab thank you so much! you mentioned in order to save the comments one need to save the draft, does it mean for this user need to have edit role but not published role?  I will give a try for the wagtail-review and see how it flows.

**Comment by zerolab** [comment]
@weiwang-gsa 
> does it mean for this user need to have edit role but not published role

not at all. if the page is in a workflow step that allows the limited user group, they can access the editor - https://github.com/wagtail/wagtail/blob/8026108fa4eae1ac384dbbb616ddda0f9c37c454/wagtail/models/__init__.py#L3997-L3998 - even if they don't have add/edit/publish permissions

**Comment by weiwang-gsa** [comment]
I just installed the wagtail-reivew extension followed the instruction in readme, but I still don't get how this would work for our scenarios.
Below are the testing steps:

1. After installation, I logged in as admin and see the Reviews showed up on left panel![Image](https://github.com/user-attachments/assets/9bb73f3d-000d-45c7-bd2e-da6d7f90069c)


2. I set up a readonly group and give its access to admin but no access to any pages. ![Image](https://github.com/user-attachments/assets/9441d715-2ebc-411a-af01-698dcc089e4b)

3. set up a workflow and add workflow task  and attach the test page to the workflow![Image](https://github.com/user-attachments/assets/1ee3d0a2-df94-4a91-a3f6-d4f9a7b92f26)
![Image](https://github.com/user-attachments/assets/50b1d04b-9893-4b4e-8b27-a8813c93ce73)

4. Created a readonly user that belong to readonly group.![Image](https://github.com/user-attachments/assets/a89e2488-d94c-43da-939b-265b2e5071a3)

5. go to the test page and put in readonly user name and submit for moderation ![Image](https://github.com/user-attachments/assets/2e8ef59e-6b8b-4811-afd1-a771ddb74a38)

6. sign out admin and logged in as readonly user, I am not able to see the "Reviews" on left hand panel, it only have "help" ![Image](https://github.com/user-attachments/assets/677b1d9b-802c-4a7d-925a-aea3484f3d12)


At this point, as a readonly user how can I see the page review or add comment?  I think maybe the editor (or admin that have full access) can send the page preview link to the readonly user so they can preview the page and add comment there.

1. So I signed back in as admin, and opened reviews page request, add some comment to the page.![Image](https://github.com/user-attachments/assets/af85fc66-4ab2-488c-811e-ba83a5244448)
2.  coppied the the page preview link and save it to my notepad
3.  Signed out of admin and signed in as read-only user, and paste the preview link just saved, well now I can see the page preview and admin's comment, but I can't add my own comment or reply admin's comment. there is no action button and it is not editable in the commenting area. ![Image](https://github.com/user-attachments/assets/8808e782-3c30-4eb4-b1a4-1419fafaf84b)


Till this step, as read-only user I just can preview the page that before publish, but still not able to put in own comment or participate in the content editing processing even I am set as the moderator for this page.  What we want is just an read-only user that can provide comments to page content and can preview the page before publish. We are not clear how we should set it up with this extension to achieve this. 

**Maybe adding a common user access "Readonly" to wagtail would be a better solution? ** for this read only user access, can just make all the function button disabled ( add, publish, edit, delete) etc, and leave "comment" function enabled, it probably would be more straight forward and intuitive solution.

**Comment by nilaygit-10721** [comment]
Ensure clear messaging or visual indicators to inform users which elements remain locked and why.

**Comment by unyimeudemy** [comment]
@zerolab I have been able to reproduce this bug and I am currently working it.

**Comment by unyimeudemy** [comment]
@zerolab  I have found the cause of the crash. The actual reason why the admin page UI crashes is because on successful login and when we try to prepare the page summary details, we get to  check what permissions the user has on different pages in PagePermissionPolicy class as shown below.

``` python
    def instances_with_direct_explore_permission(self, user):
        # Get all pages that the user has direct add/change/publish/lock permission on
        if user.is_superuser:
            # superuser has implicit permission on the root node
            return Page.objects.filter(depth=1)
        else:
            codenames = self._get_permission_codenames(
                {"add", "change", "publish", "lock"} # adding "unlock" fixes bug
            )
            return [
                perm.page
                for perm in self.get_cached_permissions_for_user(user)
                if perm.permission.codename in codenames
            ]

```
Here, we first get all the possible permissions on page before checking the ones the user have. Notice that `unlock` is not part of the set. I assume this is because the `lock` permission was suppose to do two jobs, lock and unlock ( according to official [documentation](https://docs.wagtail.org/en/stable/topics/permissions.html) )

> • **Lock** - grants the ability to lock or unlock this page (and any pages underneath it) for editing, preventing users from making any further edits to it.
> 

But this does not work because `self._get_permission_codenames({...})`  ends up returning `{'lock_page', 'change_page', 'publish_page', 'add_page'}` and when this is checked against a user that has only an unlock permission in the list comprehension, an empty array is returned.  This then causes the `instances_with_direct_explore_permission` method to return `None` to  `explorable_root_instance` method which in turn returns `None`  to `get_site_for_user` method , causing it to return a dictionary with the `root_page` key set to `None`.

```python
    return {
        "root_page": root_page,
        "root_site": root_site,
        "site_name": real_site_name if real_site_name else settings.WAGTAIL_SITE_NAME,
    }
```

This in turn causes the `get_context_data` method of the `PagesSummaryItem` class to return a dictionary with the `root_page` key set to `None`.  And this is finally sent to the `site_summary_pages.html`  where the code, `root_page.pk` , causes the admin UI to crash.

```html
<li>
    {% icon name="doc-empty" %}
    <a href="{% url 'wagtailadmin_explore' root_page.pk %}">
        {% blocktrans trimmed count counter=total_pages with total_pages|intcomma as total %}
            {{ total }} Page <span class="w-sr-only">created in {{ site_name }}</span>
        {% plural %}
            {{ total }} Pages <span class="w-sr-only">created in {{ site_name }}</span>
        {% endblocktrans %}
    </a>
</li>
```

**Path from Login to point of bug**

1. LoginView (get_success_url) →
2. HomeView (get_context_data) →
3. SiteSummaryPanel (calling hooks with name “construct_homepage_summary_items”) → PagesSummaryItem (get_context_data) → 
4. get_site_for_user → 
5. PagePermissionPolicy ( explorable_root_instance ) → 
6. PagePermissionPolicy (instances_with_direct_explore_permission).

## Proposed solutions

A direct solution will be to include “unlock” in the set and this fixes the bug. We will now have `{'unlock_page', 'change_page', 'publish_page', 'lock_page', 'add_page'}` to check against what so ever permissions the user have.

I will like to suggest the two permissions (lock and unlock) should be merged together and called something like `lock_and_unlock`, since a user shouldn’t just go about with the ability to unlock pages he/she did not lock. 

Then in other not to temper with the existing permissions (Add, Edit, Bulk delete, publish) which themselves serve a very distinct purpose, a new permission should be introduced called `View`, which specifically caters for  @weiwang-gsa  use case.

## Should I go ahead with adding  `unlock` to the set

**Comment by unyimeudemy** [comment]
Hi @gasman and @laymonage,

Could you please review my proposed solution to this issue? I would appreciate your insights to ensure I proceed in the right direction.

Thank you for your time and guidance.

**Comment by laymonage** [comment]
Thanks @unyimeudemy

> I will like to suggest the two permissions (lock and unlock) should be merged together and called something like `lock_and_unlock`, since a user shouldn’t just go about with the ability to unlock pages he/she did not lock.

That is intentional, see #5634.

> Then in other not to temper with the existing permissions (Add, Edit, Bulk delete, publish) which themselves serve a very distinct purpose, a new permission should be introduced called `View`

We can reuse Django's auto-generated `view` permission for this, but the work into adding proper support for a "read-only" permission is non-trivial, as it requires updating all the relevant places to define what's possible with a read-only permission. Not to mention that the permission system is due for a major refactoring (#2907 and wagtail/rfcs#102).

For this issue specifically, I think adding `unlock` to both of these places makes sense:
https://github.com/wagtail/wagtail/blob/784f4bbbb920c21895292e50d66659038f1acf46/wagtail/permission_policies/pages.py#L198
https://github.com/wagtail/wagtail/blob/784f4bbbb920c21895292e50d66659038f1acf46/wagtail/permission_policies/pages.py#L169

It probably also makes sense to add `bulk_delete` as well (otherwise it has the same issue), since we do allow it in a few places to check for "any" page permission:
https://github.com/wagtail/wagtail/blob/784f4bbbb920c21895292e50d66659038f1acf46/wagtail/admin/views/pages/listing.py#L269-L278
https://github.com/wagtail/wagtail/blob/784f4bbbb920c21895292e50d66659038f1acf46/wagtail/admin/auth.py#L105-L112

P.s. please note that it's the holiday season, so do not expect fast responses from people 🙂

**Comment by unyimeudemy** [comment]
@laymonage Thank you very much. I will proceed with adding `unlock` and `bulk_delete` .

**Comment by unyimeudemy** [comment]
Thank you very much, @laymonage. I will proceed with adding `unlock` and `bulk_delete`.

Do you think it would be necessary to update the [permissions documentation](https://docs.wagtail.org/en/stable/topics/permissions.html) to explicitly explain the unlock permission as outlined in [PR #5634](https://github.com/wagtail/wagtail/pull/5634)? If so, I can open a new issue to address it.

**Comment by laymonage** [comment]
Yes please. If you want to work on it you can also go ahead without creating an issue. Thanks!
